src/bifrost.cpp

```diff
@@ -0,0 +1,276 @@
+            }
+        }
+    }
+    // else, look at only position in unitig given by mapping, ignore position
+    else {
+        // iterate over specified kmer colours, update colours array with presence of those colours
+        for (it_uc; it_uc != it_uc_end; it_uc++)
+        {
+            colours_arr[it_uc.getColorID()] = 1;
+        }
+    }
+    return colours_arr;
+}
+
+// negate colour arrays
+std::vector<bool> negate_colours_array(const std::vector<bool>& array1, const std::vector<bool>& array2)
+{
+    std::vector<bool> output_array = array1;
+    for (size_t i = 0; i < array1.size(); i++)
+    {
+        if (array1[i] == 1 && array2[i] == 0)
+        {
+            output_array[i] = 0;
+        }
+    }
+    return output_array;
+}
+
+// parse a fasta and return sequences
+std::vector<std::string> parse_fasta (const std::string& fasta)
+{
+    std::vector<std::string> seq_list;
+    std::string line, DNA_sequence;
+    std::ifstream infile(fasta);
+
+    // ignore header
+    std::getline(infile, line);
+
+    // parse fasta
+    while (std::getline(infile, line)) {
+        // remove new line characters
+        line.erase(std::remove(line.begin(), line.end(), '\n'),
+                    line.end());
+
+        // erase DNA_sequence
+        DNA_sequence.clear();
+
+        // line may be empty so you *must* ignore blank lines
+        if (line.empty())
+            continue;
+
+        if (line[0] != '>') {
+            DNA_sequence = line;
+        }
+
+        // add to seq list
+        if (DNA_sequence.size() != 0)
+            seq_list.push_back(DNA_sequence);
+    }
+
+    return seq_list;
+}
+
+// query whether a unitig exists in a graph in it's entirity, if so return colour vector. If not, return empty vector.
+std::vector<bool> query_unitig (const ColoredCDBG<>& ccdbg, const std::string& query, const size_t& nb_colours)
+{
+    //split query into sequence of kmers
+    const char *query_str = query.c_str();
+
+    // find first kmer within query in graph
+    KmerIterator it_km(query_str), it_km_end;
+
+    // query the kmer in the graph, find the first unitig
+    auto unitig_current = ccdbg.find(it_km->first);
+
+    // make a const copy of the first unitig map to enable colour generation
+    const auto unitig_first = unitig_current;
+
+    // get the colours for the unitig
+    std::vector<bool> query_colours(nb_colours, 0);
+    std::vector<bool> query_colours_head;
+    std::vector<bool> query_colours_tail;
+
+//    // initialise check if kmer is first or last in sequence
+//    bool is_first = true;
+//    bool is_last = false;
+
+    // initialise unitig_found check
+    bool unitig_present = true;
+
+    // iterate to next kmer iterator
+    it_km++;
+
+    // iterate over forward strand of unitig
+    if (!unitig_current.isEmpty)
+    {
+        for (it_km; it_km != it_km_end; it_km++)
+        {
+            // query the kmer in the graph
+            auto unitig_current = ccdbg.find(it_km->first);

```

Comment:
Does this search check for the reverse complement too?

Reply:
Yes, kmers are treated canonically so in creation of a kmer object, it automatically treats the kmer and its reverse complement as the same thing.